File: examples/java/com/intel/daal/examples/decision_forest/DfClsDenseBatchModelBuilder.java
/* file: DfClsDenseBatchModelBuilder.java */
/*******************************************************************************
* Copyright 2014-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of decision forest classification model building.
 //
 //     The program builds the decision forest classification model
 //     via Model Builder and computes classification for the test data.
 ////////////////////////////////////////////////////////////////////////////////
 */

/**
 * <a name="DAAL-EXAMPLE-JAVA-DfClsDenseBatchModelBuilder">
 * @example DfClsDenseBatchModelBuilder.java
 */

package com.intel.daal.examples.decision_forest;

import com.intel.daal.algorithms.classifier.prediction.ModelInputId;
import com.intel.daal.algorithms.classifier.prediction.NumericTableInputId;
import com.intel.daal.algorithms.classifier.prediction.PredictionResult;
import com.intel.daal.algorithms.classifier.prediction.PredictionResultId;
import com.intel.daal.algorithms.decision_forest.classification.Model;
import com.intel.daal.algorithms.decision_forest.classification.ModelBuilder;
import com.intel.daal.algorithms.decision_forest.classification.prediction.*;
import com.intel.daal.algorithms.decision_forest.*;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.MergedNumericTable;
import com.intel.daal.data_management.data_source.DataSource;
import com.intel.daal.data_management.data_source.FileDataSource;
import com.intel.daal.examples.utils.Service;
import com.intel.daal.services.DaalContext;
import com.intel.daal.data_management.data.*;

class DfClsDenseBatchModelBuilder {
    /* Input data set parameters */
    private static final String testDatasetFileName = "../data/batch/df_classification_model_builder_test.csv";

    private static final long categoricalFeaturesIndices [] =  { 2 };

    /* Number of features in training and testing data sets */
    private static final int nFeatures     = 3;
    /* Number of classes */
    private static final int nClasses      = 5;
    /* Number of tree in decision forest classification model */
    private static final int nTrees = 3;
    private static final int minObservationsInLeafNode = 8;

    private static NumericTable testGroundTruth;
    static PredictionResult predictionResult;
    static Model model;

    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        buildModel();
        testModel();
        printResults(predictionResult);

        context.dispose();
    }

    public static void buildModel() {
        final long nNodes = 3;

        ModelBuilder modelBuilder = new ModelBuilder(context, nClasses, nTrees);

        long tree1 = modelBuilder.createTree(nNodes);
        long root1   = modelBuilder.addSplitNode(tree1, ModelBuilder.noParent, 0, 0, 0.174108);
        long child11 = modelBuilder.addLeafNode(tree1, root1, 0, 0);
        long child12 = modelBuilder.addLeafNode(tree1, root1, 1, 4);

        long tree2 = modelBuilder.createTree(nNodes);
        long root2 = modelBuilder.addSplitNode(tree2, modelBuilder.noParent, 0, 1, 0.571184);
        long child22 = modelBuilder.addLeafNode(tree2, root2, 1, 4);
        long child21 = modelBuilder.addLeafNode(tree2, root2, 0, 2);
        long tree3 = modelBuilder.createTree(nNodes);
        long root3 = modelBuilder.addSplitNode(tree3, modelBuilder.noParent, 0, 0, 0.303995);
        long child32 = modelBuilder.addLeafNode(tree3, root3, 1, 4);
        long child31 = modelBuilder.addLeafNode(tree3, root3, 0, 2);

        model = modelBuilder.getModel();
    }

    private static void testModel() {

        /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
        FileDataSource testDataSource = new FileDataSource(context, testDatasetFileName,
                DataSource.DictionaryCreationFlag.DoDictionaryFromContext,
                DataSource.NumericTableAllocationFlag.NotAllocateNumericTable);

        /* Create Numeric Tables for testing data and labels */
        NumericTable testData = new HomogenNumericTable(context, Float.class, nFeatures, 0, NumericTable.AllocationFlag.NotAllocate);
        testGroundTruth = new HomogenNumericTable(context, Float.class, 1, 0, NumericTable.AllocationFlag.NotAllocate);
        MergedNumericTable mergedData = new MergedNumericTable(context);
        mergedData.addNumericTable(testData);
        mergedData.addNumericTable(testGroundTruth);

        /* Retrieve the data from an input file */
        testDataSource.loadDataBlock(mergedData);
        /* Set feature as categorical */
        testData.getDictionary().setFeature(Float.class, 2, DataFeatureUtils.FeatureType.DAAL_CATEGORICAL);

        /* Create algorithm objects for decision forest classification prediction with the fast method */
        PredictionBatch algorithm = new PredictionBatch(context, Float.class, PredictionMethod.defaultDense, nClasses);

        /* Pass a testing data set and the trained model to the algorithm */
        algorithm.input.set(NumericTableInputId.data, testData);
        algorithm.input.set(ModelInputId.model, model);

        /* Compute prediction results */
        predictionResult = algorithm.compute();
    }

    private static void printResults(PredictionResult predictionResult) {
        NumericTable predictionResults = predictionResult.get(PredictionResultId.prediction);
        Service.printNumericTable("Decision forest prediction results (first 10 rows):", predictionResults, 10);
        Service.printNumericTable("Ground truth (first 10 rows):", testGroundTruth, 10);
    }

}

File: examples/daal/cpp/source/decision_forest/df_cls_dense_batch_model_builder.cpp
/* file: df_cls_dense_batch_model_builder.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of decision forest classification model building.
!
!    The program builds the decision forest classification model
!     via Model Builder and computes classification for the test data.
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-DF_CLS_MODEL_BUILDER"></a>
 * \example df_cls_dense_batch_model_builder.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::algorithms;
using namespace daal::data_management;
using namespace daal::algorithms::decision_forest::classification;

/* Input data set parameters */
const std::string testDatasetFileName = "../data/batch/df_classification_model_builder_test.csv";
const size_t categoricalFeaturesIndices[] = { 2 };
const size_t nFeatures = 3; /* Number of features in training and testing data sets */

/* Decision forest parameters */
const size_t nTrees = 3;
const size_t nClasses = 5; /* Number of classes */

void testModel(decision_forest::classification::ModelPtr& model);
decision_forest::classification::ModelPtr buildModel();
void loadData(const std::string& fileName, NumericTablePtr& pData, NumericTablePtr& pDependentVar);

int main(int argc, char* argv[]) {
    checkArguments(argc, argv, 1, &testDatasetFileName);

    decision_forest::classification::ModelPtr model = buildModel();
    testModel(model);

    return 0;
}

decision_forest::classification::ModelPtr buildModel() {
    const size_t nNodes = 3;
    const int defaultLeft = 0;
    const double cover = 0.0;

    ModelBuilder modelBuilder(nClasses, nTrees);
    ModelBuilder::TreeId tree1 = modelBuilder.createTree(nNodes);
    ModelBuilder::NodeId root1 =
        modelBuilder.addSplitNode(tree1, ModelBuilder::noParent, 0, 0, 0.174108, defaultLeft, cover);
    /* ModelBuilder::NodeId child12 = */ modelBuilder.addLeafNode(tree1, root1, 1, 4, cover);
    double proba11[] = { 0.8, 0.1, 0.0, 0.1, 0.0 };
    /* ModelBuilder::NodeId child11 = */ modelBuilder.addLeafNodeByProba(tree1, root1, 0, proba11, cover);

    ModelBuilder::TreeId tree2 = modelBuilder.createTree(nNodes);
    ModelBuilder::NodeId root2 =
        modelBuilder.addSplitNode(tree2, ModelBuilder::noParent, 0, 1, 0.571184, defaultLeft, cover);
    /* ModelBuilder::NodeId child22 = */ modelBuilder.addLeafNode(tree2, root2, 1, 4, cover);
    /* ModelBuilder::NodeId child21 = */ modelBuilder.addLeafNode(tree2, root2, 0, 2, cover);

    ModelBuilder::TreeId tree3 = modelBuilder.createTree(nNodes);
    ModelBuilder::NodeId root3 =
        modelBuilder.addSplitNode(tree3, ModelBuilder::noParent, 0, 0, 0.303995, defaultLeft, cover);
    double proba32[] = { 0.05, 0.1, 0.0, 0.1, 0.75 };
    /* ModelBuilder::NodeId child32 = */ modelBuilder.addLeafNodeByProba(tree3, root3, 1, proba32, cover);
    /* ModelBuilder::NodeId child31 = */ modelBuilder.addLeafNode(tree3, root3, 0, 2, cover);
    modelBuilder.setNFeatures(nFeatures);
    return modelBuilder.getModel();
}

void testModel(decision_forest::classification::ModelPtr& model) {
    /* Create Numeric Tables for testing data and ground truth values */
    NumericTablePtr testData;
    NumericTablePtr testGroundTruth;

    loadData(testDatasetFileName, testData, testGroundTruth);

    /* Create an algorithm object to predict values of decision forest classification */
    prediction::Batch<> algorithm(nClasses);

    /* Pass a testing data set and the trained model to the algorithm */
    algorithm.input.set(classifier::prediction::data, testData);

    /* set model obtained by builder */
    algorithm.input.set(classifier::prediction::model, model);

    /* set voting method */
    algorithm.parameter().votingMethod = prediction::unweighted;

    /* Predict values of decision forest classification */
    algorithm.compute();

    /* Retrieve the algorithm results */
    classifier::prediction::ResultPtr predictionResult = algorithm.getResult();
    printNumericTable(predictionResult->get(classifier::prediction::prediction),
                      "Decision forest prediction results (first 10 rows):",
                      10);
    printNumericTable(testGroundTruth, "Ground truth (first 10 rows):", 10);
}

void loadData(const std::string& fileName, NumericTablePtr& pData, NumericTablePtr& pDependentVar) {
    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> trainDataSource(fileName,
                                                      DataSource::notAllocateNumericTable,
                                                      DataSource::doDictionaryFromContext);

    /* Create Numeric Tables for training data and dependent variables */
    pData.reset(new HomogenNumericTable<>(nFeatures, 0, NumericTable::notAllocate));
    pDependentVar.reset(new HomogenNumericTable<>(1, 0, NumericTable::notAllocate));
    NumericTablePtr mergedData(new MergedNumericTable(pData, pDependentVar));

    /* Retrieve the data from input file */
    trainDataSource.loadDataBlock(mergedData.get());

    NumericTableDictionaryPtr pDictionary = pData->getDictionarySharedPtr();
    for (size_t i = 0,
                n = sizeof(categoricalFeaturesIndices) / sizeof(categoricalFeaturesIndices[0]);
         i < n;
         ++i)
        (*pDictionary)[categoricalFeaturesIndices[i]].featureType =
            data_feature_utils::DAAL_CATEGORICAL;
}
